### ISSUE TITLE
False Negative: UnnecessaryParenthesesCheck

### ISSUE THREAD

**Issue by Vyom-Yadav** [checkstyle/checkstyle#11109]
I have read check documentation: https://checkstyle.sourceforge.io/config_coding.html#UnnecessaryParentheses
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ javac Test.java

/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="UnnecessaryParentheses"/>
    </module>
</module>

/var/tmp $ cat Test.java
public class Test {
    void method() {
        boolean a = true;
        boolean b = false;
        if(a || (b ^ a) || b) { // expected violation
        }
        if(a || (b | a) || b) { // expected violation
        }
        if(a || (b & a) || b) { // expected violation
        }
        int c = 1;
        int d = 12;
        if(a || (c ^ d) >> 1 == 1 || b) { // ok
        }
    }
}
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-9.2-all.jar -c config.xml Test.java
Starting audit...
Audit done.
```

---

**Describe what you expect in detail.**
```
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-9.2-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/vyom/IdeaProjects/ConfigCheckstyle/src/TestP/Test.java:5:17: Unnecessary parentheses around expression. [UnnecessaryParentheses]
[ERROR] /home/vyom/IdeaProjects/ConfigCheckstyle/src/TestP/Test.java:7:17: Unnecessary parentheses around expression. [UnnecessaryParentheses]
[ERROR] /home/vyom/IdeaProjects/ConfigCheckstyle/src/TestP/Test.java:9:17: Unnecessary parentheses around expression. [UnnecessaryParentheses]
Audit done.
Checkstyle ends with 3 errors.

```
Currently, the check does not identify any of `|`, `&` and `^`. Though we have had reports that UnnecessaryParentheses is too strict ([example](https://github.com/checkstyle/checkstyle/issues/10946)) so it would be better to discuss whether adding support for it will be worth it.

**Comment by rnveach** [comment]
Issue looks valid to me.

**Comment by nmancus1** [comment]
> Currently, the check does not identify any of `|`, `&` and `^`

If all three tokens are part of this false negative issue, please add them to your example.

From the check documentation, we should keep this in mind while evaluating issue: 

> It [UnnecessaryParenthesesCheck] also doesn't know about operator precedence and associativity

**Comment by Vyom-Yadav** [comment]
> If all three tokens are part of this false negative issue, please add them to your example.

Added.

>From the check documentation, we should keep this in mind while evaluating issue:
>
>>It [UnnecessaryParenthesesCheck] also doesn't know about operator precedence and associativity

This is partially true, we have started adding some support for operator precedence now, check can now determine unnecessary parentheses around a mix of `||`, `&&` being used with relational, unary and postfix operators, though as the check was not type aware we had to drop some cases.

**Comment by Vyom-Yadav** [comment]
@rnveach @nmancus1 Approved tag?

**Comment by nmancus1** [comment]
@Vyom-Yadav wouldn't this require for the check to be type aware? If so, how do we handle case where operand is declared elsewhere (not in current compilation unit)?

**Comment by romani** [comment]
I do not think we need type for this.
expression parts should be boolean and this is enough for us.

I think we should approve this issue.
@nmancus1 , as you only who have concerns, please put label as you resolve your concerns.

**Comment by nmancus1** [comment]
@romani approved, I am not sure if we should consider this a new feature or bug, I think new feature is probably the best fit.

**Comment by romani** [comment]
We approve issue first.
After implantation is done and tested we define new label that is based 100% on implementation details.
Issue can looks simple bug but results in breaking compatibility.

**Comment by rnveach** [comment]
Fix was merged
